packages/dd-trace/src/datastreams/processor.js

```diff
@@ -0,0 +1,193 @@
+const pkg = require('../../../../package.json')
+// Message pack int encoding is done in big endian, but data streams uses little endian
+const Uint64 = require('int64-buffer').Uint64BE
+
+const { LogCollapsingLowestDenseDDSketch } = require('@datadog/sketches-js')
+
+const { DataStreamsWriter } = require('./writer')
+const { computePathwayHash } = require('./pathway')
+const ENTRY_PARENT_HASH = Buffer.from('0000000000000000', 'hex')
+
+const HIGH_ACCURACY_DISTRIBUTION = 0.0075
+
+class StatsPoint {
+  constructor (hash, parentHash, edgeTags) {
+    this.hash = new Uint64(hash)
+    this.parentHash = new Uint64(parentHash)
+    this.edgeTags = edgeTags
+    this.edgeLatency = new LogCollapsingLowestDenseDDSketch(HIGH_ACCURACY_DISTRIBUTION)
+    this.pathwayLatency = new LogCollapsingLowestDenseDDSketch(HIGH_ACCURACY_DISTRIBUTION)
+  }
+
+  addLatencies (checkpoint) {
+    const edgeLatencySec = checkpoint.edgeLatencyNs / 1e9
+    const pathwayLatencySec = checkpoint.pathwayLatencyNs / 1e9
+    this.edgeLatency.accept(edgeLatencySec)
+    this.pathwayLatency.accept(pathwayLatencySec)
+  }
+
+  encode () {
+    return {
+      Hash: this.hash,
+      ParentHash: this.parentHash,
+      EdgeTags: this.edgeTags,
+      EdgeLatency: this.edgeLatency.toProto(),
+      PathwayLatency: this.pathwayLatency.toProto()
+    }
+  }
+}
+
+class StatsBucket extends Map {
+  forCheckpoint (checkpoint) {
+    const key = checkpoint.hash
+    if (!this.has(key)) {
+      this.set(key, new StatsPoint(checkpoint.hash, checkpoint.parentHash, checkpoint.edgeTags)) // StatsPoint
+    }
+
+    return this.get(key)
+  }
+}
+
+class TimeBuckets extends Map {
+  forTime (time) {
+    if (!this.has(time)) {
+      this.set(time, new StatsBucket())
+    }
+
+    return this.get(time)
+  }
+}
+
+class DataStreamsProcessor {
+  constructor ({
+    dsmEnabled,
+    hostname,
+    port,
+    url,
+    env,
+    tags
+  } = {}) {
+    this.writer = new DataStreamsWriter({
+      hostname,
+      port,
+      url
+    })
+    this.bucketSizeNs = 1e10
+    this.buckets = new TimeBuckets()
+    this.hostname = os.hostname()
+    this.enabled = dsmEnabled
+    this.env = env
+    this.tags = tags || {}
+    this.service = this.tags.service || 'unnamed-nodejs-service'
+    this.sequence = 0
+
+    if (this.enabled) {
+      this.timer = setInterval(this.onInterval.bind(this), 10000)
+      this.timer.unref()
+    }
+  }
+
+  onInterval () {
+    const serialized = this._serializeBuckets()
+    if (!serialized) return
+    const payload = {
+      Env: this.env,
+      Service: this.service,
+      Stats: serialized,
+      TracerVersion: pkg.version,
+      Lang: 'javascript'
+    }
+    this.writer.flush(payload)
+  }
+
+  recordCheckpoint (checkpoint) {
+    if (!this.enabled) return
+    const bucketTime = Math.round(checkpoint.currentTimestamp - (checkpoint.currentTimestamp % this.bucketSizeNs))
+    this.buckets.forTime(bucketTime)
+      .forCheckpoint(checkpoint)
```

Comment:
Is this actually by checkpoint? This seems fairly expensive, unless we don't expect a lot of checkpoints?

Reply:
Are you asking if this is run once for every checkpoint in a chain of services? Like `svc1 -> svc2 -> svc3`? If so, this isn't the case, and the code isn't iterating for each of the prior services.

Reply:
I think the function is better named as "forThisCheckpoint"